From a7e9101b4409cee1cab24552e2c94298e3b73463 Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Wed, 27 Jul 2022 17:51:00 -0700
Subject: [PATCH 10/11] HACK: LONGTERM: Revert "powerpc/44x: Fix build failure
 with GCC 12 (unrecognized opcode: `wrteei')"

This reverts commit 0d913358a816b35fa5b75385e80f63b9abcf1b27.

Link: https://github.com/ClangBuiltLinux/linux/issues/1679
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 arch/powerpc/include/asm/barrier.h        |  2 --
 arch/powerpc/include/asm/nohash/pgtable.h |  2 +-
 arch/powerpc/include/asm/synch.h          |  5 +----
 arch/powerpc/mm/nohash/tlb_low.S          |  4 ++--
 arch/powerpc/platforms/Kconfig.cputype    | 19 +------------------
 arch/powerpc/sysdev/fsl_rio.c             | 12 ++----------
 6 files changed, 7 insertions(+), 37 deletions(-)

diff --git a/arch/powerpc/include/asm/barrier.h b/arch/powerpc/include/asm/barrier.h
index b95b666f0374..2e0a443578c7 100644
--- a/arch/powerpc/include/asm/barrier.h
+++ b/arch/powerpc/include/asm/barrier.h
@@ -42,8 +42,6 @@
 /* The sub-arch has lwsync */
 #if defined(CONFIG_PPC64) || defined(CONFIG_PPC_E500MC)
 #    define SMPWMB      LWSYNC
-#elif defined(CONFIG_BOOKE)
-#    define SMPWMB      mbar
 #else
 #    define SMPWMB      eieio
 #endif
diff --git a/arch/powerpc/include/asm/nohash/pgtable.h b/arch/powerpc/include/asm/nohash/pgtable.h
index a6caaaab6f92..53a3e48a0695 100644
--- a/arch/powerpc/include/asm/nohash/pgtable.h
+++ b/arch/powerpc/include/asm/nohash/pgtable.h
@@ -188,7 +188,7 @@ static inline void __set_pte_at(struct mm_struct *mm, unsigned long addr,
 	if (IS_ENABLED(CONFIG_PPC32) && IS_ENABLED(CONFIG_PTE_64BIT) && !percpu) {
 		__asm__ __volatile__("\
 			stw%X0 %2,%0\n\
-			mbar\n\
+			eieio\n\
 			stw%X1 %L2,%1"
 		: "=m" (*ptep), "=m" (*((unsigned char *)ptep+4))
 		: "r" (pte) : "memory");
diff --git a/arch/powerpc/include/asm/synch.h b/arch/powerpc/include/asm/synch.h
index b0b4c64870d7..de740dff4f65 100644
--- a/arch/powerpc/include/asm/synch.h
+++ b/arch/powerpc/include/asm/synch.h
@@ -14,10 +14,7 @@ extern void do_lwsync_fixups(unsigned long value, void *fixup_start,
 
 static inline void eieio(void)
 {
-	if (IS_ENABLED(CONFIG_BOOKE))
-		__asm__ __volatile__ ("mbar" : : : "memory");
-	else
-		__asm__ __volatile__ ("eieio" : : : "memory");
+	__asm__ __volatile__ ("eieio" : : : "memory");
 }
 
 static inline void isync(void)
diff --git a/arch/powerpc/mm/nohash/tlb_low.S b/arch/powerpc/mm/nohash/tlb_low.S
index e1199608ff4d..4a0ea62aa2d5 100644
--- a/arch/powerpc/mm/nohash/tlb_low.S
+++ b/arch/powerpc/mm/nohash/tlb_low.S
@@ -186,7 +186,7 @@ _GLOBAL(_tlbivax_bcast)
 	isync
 	PPC_TLBIVAX(0, R3)
 	isync
-	mbar
+	eieio
 	tlbsync
 BEGIN_FTR_SECTION
 	b	1f
@@ -355,7 +355,7 @@ _GLOBAL(_tlbivax_bcast)
 	rlwimi	r4,r6,MAS6_SIND_SHIFT,MAS6_SIND
 1:	mtspr	SPRN_MAS6,r4		/* assume AS=0 for now */
 	PPC_TLBIVAX(0,R3)
-	mbar
+	eieio
 	tlbsync
 	sync
 	wrtee	r10
diff --git a/arch/powerpc/platforms/Kconfig.cputype b/arch/powerpc/platforms/Kconfig.cputype
index 046b571496b1..ba7c58a3f107 100644
--- a/arch/powerpc/platforms/Kconfig.cputype
+++ b/arch/powerpc/platforms/Kconfig.cputype
@@ -138,7 +138,7 @@ config POWERPC64_CPU
 
 config POWERPC_CPU
 	bool "Generic 32 bits powerpc"
-	depends on PPC_BOOK3S_32
+	depends on PPC32 && !PPC_8xx && !PPC_85xx && !40x
 
 config CELL_CPU
 	bool "Cell Broadband Engine"
@@ -196,20 +196,6 @@ config 405_CPU
 	depends on 40x
 	depends on !CC_IS_CLANG
 
-config 440_CPU
-	bool "440 (44x family)"
-	depends on 44x
-
-config 464_CPU
-	bool "464 (44x family)"
-	depends on 44x
-	depends on !CC_IS_CLANG
-
-config 476_CPU
-	bool "476 (47x family)"
-	depends on PPC_47x
-	depends on !CC_IS_CLANG
-
 config 860_CPU
 	bool "8xx family"
 	depends on PPC_8xx
@@ -262,9 +248,6 @@ config TARGET_CPU
 	default "power4" if POWERPC64_CPU && !CPU_LITTLE_ENDIAN
 	default "power8" if POWERPC64_CPU && CPU_LITTLE_ENDIAN
 	default "405" if 405_CPU
-	default "440" if 440_CPU
-	default "464" if 464_CPU
-	default "476" if 476_CPU
 	default "860" if 860_CPU
 	default "e300c2" if E300C2_CPU
 	default "e300c3" if E300C3_CPU
diff --git a/arch/powerpc/sysdev/fsl_rio.c b/arch/powerpc/sysdev/fsl_rio.c
index c8f044d62fe2..272762b937ac 100644
--- a/arch/powerpc/sysdev/fsl_rio.c
+++ b/arch/powerpc/sysdev/fsl_rio.c
@@ -69,10 +69,10 @@
 
 static DEFINE_SPINLOCK(fsl_rio_config_lock);
 
-#define ___fsl_read_rio_config(x, addr, err, op, barrier)	\
+#define __fsl_read_rio_config(x, addr, err, op)		\
 	__asm__ __volatile__(				\
 		"1:	"op" %1,0(%2)\n"		\
-		"	"barrier"\n"			\
+		"	eieio\n"			\
 		"2:\n"					\
 		".section .fixup,\"ax\"\n"		\
 		"3:	li %1,-1\n"			\
@@ -83,14 +83,6 @@ static DEFINE_SPINLOCK(fsl_rio_config_lock);
 		: "=r" (err), "=r" (x)			\
 		: "b" (addr), "i" (-EFAULT), "0" (err))
 
-#ifdef CONFIG_BOOKE
-#define __fsl_read_rio_config(x, addr, err, op)	\
-	___fsl_read_rio_config(x, addr, err, op, "mbar")
-#else
-#define __fsl_read_rio_config(x, addr, err, op)	\
-	___fsl_read_rio_config(x, addr, err, op, "eieio")
-#endif
-
 void __iomem *rio_regs_win;
 void __iomem *rmu_regs_win;
 resource_size_t rio_law_start;
-- 
2.39.2

